designs/simplified-scaffolding.md

```diff
@@ -0,0 +1,512 @@
+
+This works fine when you have a single "major" Kind, but quickly grows
+unweildy when you have multiple major kinds and end up with
+a hundreds-of-lines-long `types.go` file (e.g. the `appsv1` API group in
+core Kubernetes).  Splitting out by "major" Kind (`Deployment`,
+`ReplicaSet`, etc) makes the code organization clearer.
+
+#### Change the current scaffold to just make Add a method on the reconciler
+
+While this solves the dependency issues (mostly, since you might want to
+further pass configuration to the setup logic and not just the runtime
+logic), it does not solve the underlying pedagogical issues around the
+initial structure burying key logic amidst a sprawl of generated files and
+directories.
+
+### Making this work with multiple controllers, API versions, API groups, etc
+
+#### Versions
+
+Most projects will eventually grow multiple API versions.  The only
+wrinkle here is making sure API versions get added to a scheme.  This can
+be solved by adding a specially-marked init function that registration
+functions get added to (see the example).
+
+#### Groups
+
+Some projects eventually grow multiple API groups.  Presumably, in the
+case of multiple API groups, the desired hierarchy is:
+
+```shell
+$ tree ./test/project/api
+./test/project/api
+├── groupa
+│   └── v1
+│       └── types.go
+└── groupb
+    └── v1
+        └── types.go
+```
+
+There are three options here:
+
+1. Scaffold with the more complex API structure (this looks pretty close
+   to what we do today).  It doesn't add a ton of complexity, but does
+   bury types deeper in a directory structure.
+
+2. Try to move things and rename references.  This takes a lot more effort
+   on the KubeBuilder maintainers' part if we try to rename references
+   across the codebase.  Not so much if we force the user to, but that's
+   a poorer experience.
+
+3. Tell users to move things, and scaffold out with the new structure.
+   This is fairly messy for the user.
+
+Since growing to multiple API groups seems to be fairly uncommon, it's
+mostly like safe to take a hybrid approach here -- allow manually
+specifying the output path, and, when not specified, asking the user to
+first restructure before running the command.
+
+#### Controllers
+
+Multiple controllers don't need their own package, but we'd want to
+scaffold out the builder.  We have two options here:
+
+1. Looking for a particular code comment, and appending a new builder
+   after it.  This is a bit more complicated for us, but perhaps provides
+   a nicer UX.
+
+2. Simply adding a new controller, and reminding the user to add the
+   builder themselves.  This is easier for the maintainers, but perhaps
+   a slighly poorer UX for the users.  However, writing out a builder by
+   hand is significantly less complex than adding a controller by hand in
+   the current structure.
+
+Option 1 should be fairly simple, since the logic is already needed for
+registering types to the scheme.
```

Comment:
In case where user has removed that particular code comment (or we are unable to wire second controller for some reason), we can always fallback to option 2 where we just emit the snippet which user can copy-paste in their main.go.